File: server/src/main/java/model/Currency.java
package model;

import manager.LanguageUtility;
import manager.Utility;
import org.apache.commons.configuration2.Configuration;

import java.util.ArrayList;
import java.util.List;

import static manager.Utility.consumeNumber;
import static manager.Utility.consumeString;

/**
 * The coin value is defined in copper coins
 */
public class Currency implements ICurrency {

    protected final int coinValue;
    protected final String coinString;
    protected final boolean tradeable;

    /**
     * Represents a not tradeable currency object
     */
    public Currency() {
        this.tradeable = false;
        this.coinValue = 0;
        this.coinString = LanguageUtility.getMessage("coin.notTradeable");
    }

    /**
     * Creates a currency object with a predefined number of coins
     *
     * @param coinValue amount of  copper coins
     */
    public Currency(int coinValue) {
        this.tradeable = true;
        this.coinValue = coinValue;
        this.coinString = toCoinString(coinValue);
    }

    /**
     * Creates a currency object which matches the value
     * defined in the coinString
     *
     * @param coinString in the format of 7G 3S 11K
     */
    public Currency(String coinString) {
        if (coinString.equalsIgnoreCase(LanguageUtility.getMessage("coin.notTradeable"))) {
            this.tradeable = false;
            this.coinValue = 0;
            this.coinString = LanguageUtility.getMessage("coin.notTradeable");
        } else {
            this.tradeable = true;
            this.coinString = coinString;
            this.coinValue = toCoinValue(coinString);
        }
    }

    public Currency(int copper, int silver, int gold) {
        this.tradeable = true;
        Configuration config = Utility.getConfig();
        int silverToCopper = config.getInt("coin.silver.toCopper");
        int goldToCopper = config.getInt("coin.gold.toSilver") * silverToCopper;

        this.coinValue = copper + silverToCopper * silver + goldToCopper * gold;
        this.coinString = toCoinString(coinValue);
    }

    @Override
    public ICurrency add(ICurrency other) {
        return new Currency(getCoinValue() + other.getCoinValue());
    }

    @Override
    public ICurrency add(int value) {
        return new Currency(getCoinValue() + value);
    }

    @Override
    public ICurrency sub(ICurrency other) {
        return new Currency(getCoinValue() - other.getCoinValue());
    }

    @Override
    public ICurrency sub(int value) {
        return new Currency(getCoinValue() - value);
    }

    @Override
    public ICurrency multiply(float multiplicative) {
        return new Currency(Math.round(getCoinValue() * multiplicative));
    }

    @Override
    public ICurrency divide(float d) {
        return new Currency(Math.round(getCoinValue() / d));
    }

    /**
     * Reads the cost from the String and
     * returns is a a amount of copper coins
     *
     * @param cost String to read
     * @return amount of copper coins
     */
    protected int toCoinValue(String cost) {
        if (cost.isBlank()) {
            return 0;
        }
        Configuration config = Utility.getConfig();

        int value = 0;
        int silverToCopper = config.getInt("coin.silver.toCopper");
        int goldToCopper = config.getInt("coin.gold.toSilver") * silverToCopper;
        String copper = LanguageUtility.getMessage("coin.copper.short");
        String silver = LanguageUtility.getMessage("coin.silver.short");
        String gold = LanguageUtility.getMessage("coin.gold.short");

        List<Character> costList = new ArrayList<>();
        for (char c : cost.toCharArray()) {
            costList.add(c);
        }

        while (!costList.isEmpty()) {
            int amount = consumeNumber(costList);
            String coin = consumeString(costList);

            if (coin.equals(copper)) {
                value += amount;
            } else if (coin.equals(silver)) {
                value += amount * silverToCopper;
            } else if (coin.equals(gold)) {
                value += amount * goldToCopper;
            }
        }

        return value;
    }

    /**
     * Visualises the amount of copper coins
     * in copper, silver and gold
     *
     * @param cost amount of copper coins
     * @return human-readable format
     */
    protected static String toCoinString(int cost) {
        int silverToCopper = Utility.getConfig().getInt("coin.silver.toCopper");
        int goldToSilver = Utility.getConfig().getInt("coin.gold.toSilver");
        String copperCoin = LanguageUtility.getMessage("coin.copper.short");
        String silverCoin = LanguageUtility.getMessage("coin.silver.short");
        String goldCoin = LanguageUtility.getMessage("coin.gold.short");

        int copper = cost % silverToCopper;
        cost /= silverToCopper;
        int silver = cost % goldToSilver;
        cost /= goldToSilver;
        int gold = cost;

        StringBuilder result = new StringBuilder();

        if (gold > 0) {
            result.append(gold).append(goldCoin).append(" ");
        }
        if (silver > 0) {
            result.append(silver).append(silverCoin).append(" ");
        }
        if (copper > 0) {
            result.append(copper).append(copperCoin);
        }

        if (result.length() == 0) {
            return "0" + copperCoin;
        }

        return result.toString().trim();
    }

    /**
     * The amount of copper coins this currency object represents
     */
    public int getCoinValue() {
        return coinValue;
    }

    /**
     * A human readable String of the value that this object represents
     */
    public String getCoinString() {
        return coinString;
    }

    /**
     * If this currency object represents a tradeable value
     */
    public boolean isTradeable() {
        return tradeable;
    }

    @Override
    public String toString() {
        if (tradeable) {
            return getCoinString();
        } else {
            return LanguageUtility.getMessage("coin.notTradeable");
        }
    }
}

File: app/src/main/java/de/pnp/manager/app/model/Currency.kt
package de.pnp.manager.app.model

import de.pnp.manager.app.state.ConfigHelper
import de.pnp.manager.model.ICurrency

class Currency() : ICurrency {

    private var coinValue = 0
    private var coinString: String = ""
    protected var tradeable = false

    constructor(valueAsString: String): this() {
        if (valueAsString.isBlank() || valueAsString.equals("Nicht Handelbar",true)) {
            return
        }

        var value = 0
        val silverToCopper: Int = ConfigHelper.getInt("coin.silver.toCopper")
        val goldToCopper: Int = ConfigHelper.getInt("coin.gold.toSilver") * silverToCopper
        val copper = "K"//val copper: String = LanguageUtility.getMessage("coin.copper.short")
        val silver = "S" //val silver: String = LanguageUtility.getMessage("coin.silver.short")
        val gold = "G"//val gold: String = LanguageUtility.getMessage("coin.gold.short")

        val costList: ArrayList<Char> = ArrayList()
        for (c in valueAsString.trim().toCharArray()) {
            costList.add(c)
        }

        while (costList.isNotEmpty()) {
            val amount: Int = consumeNumber(costList)
            val coin: String = consumeString(costList)
            if (coin == copper) {
                value += amount
            } else if (coin == silver) {
                value += amount * silverToCopper
            } else if (coin == gold) {
                value += amount * goldToCopper
            }
        }

        coinValue = value
        coinString = valueAsString
    }

    constructor(copperValue : Int) : this() {
        coinValue = copperValue
    }

    init {
        this.tradeable = false
        this.coinValue = 0
        this.coinString = "" //TODO LanguageUtility.getMessage("coin.notTradeable")
    }

    override fun add(other: ICurrency?): ICurrency {
        return Currency(getCoinValue() + other!!.coinValue)
    }

    override fun add(value: Int): ICurrency {
        return Currency(getCoinValue() + value)
    }

    override fun sub(other: ICurrency?): ICurrency {
        return Currency(getCoinValue() - other!!.coinValue)
    }

    override fun sub(value: Int): ICurrency {
        return Currency(getCoinValue() - value)
    }

    override fun multiply(multiplicative: Float): ICurrency {
        return Currency(Math.round(getCoinValue() * multiplicative))
    }

    override fun divide(d: Float): ICurrency {
        return Currency(Math.round(getCoinValue() / d))
    }

    fun getGoldAmount() : Int {
        val silverToCopper: Int = ConfigHelper.getInt("coin.silver.toCopper")
        val goldToSilver: Int = ConfigHelper.getInt("coin.gold.toSilver")

        var cost = coinValue
        val copper: Int = cost % silverToCopper
        cost /= silverToCopper
        val silver: Int = cost % goldToSilver
        cost /= goldToSilver
        val gold: Int = cost

        println("gold: " + gold)
        return gold
    }

    fun getSilverAmount(): Int {
        val silverToCopper: Int = ConfigHelper.getInt("coin.silver.toCopper")
        val goldToSilver: Int = ConfigHelper.getInt("coin.gold.toSilver")

        var cost = coinValue
        val copper: Int = cost % silverToCopper
        cost /= silverToCopper

        println("silver: " + cost % goldToSilver)
        return cost % goldToSilver
    }

    fun getCopperAmount(): Int {
        val silverToCopper: Int = ConfigHelper.getInt("coin.silver.toCopper")
        val goldToSilver: Int = ConfigHelper.getInt("coin.gold.toSilver")

        val cost = coinValue

        println("copper: " + cost % silverToCopper)
        return cost % silverToCopper
    }

    override fun getCoinValue(): Int {
        return this.coinValue
    }

    override fun getCoinString(): String {
        TODO("Not yet implemented")
    }

    override fun isTradeable(): Boolean {
        TODO("Not yet implemented")
    }

    /**
     * Reads and consumes all chars from the input list
     * until if finds a non digit character
     *
     * @param input list that gets consumed
     * @return the parsed digits as int
     */
    private fun consumeNumber(input: ArrayList<Char>): Int {
        val number = StringBuilder()
        while (!input.isEmpty()) {
            val c = input[0]
            if (Character.isDigit(c)) {
                input.removeAt(0)
                number.append(c)
            } else {
                break
            }
        }
        return try {
            number.toString().toInt()
        } catch (e: NumberFormatException) {
            0
        }
    }

    /**
     * Reads and consumes all chars from the input list
     * until if finds a digit character
     *
     * @param input list that gets consumed
     * @return the parsed chars as trimmed String
     */
    private fun consumeString(input: ArrayList<Char>): String {
        val result = StringBuilder()
        while (!input.isEmpty()) {
            val c = input[0]
            if (!Character.isDigit(c)) {
                input.removeAt(0)
                result.append(c)
            } else {
                break
            }
        }
        return result.toString().trim { it <= ' ' }
    }
}